### PR TITLE
[WIP] Use BinaryBuilder for all available dependencies by default

### PR DESCRIPTION
**Pull request by staticfloat**
The list of BB-enabled build dependencies is now:

- LLVM
- OpenBLAS
- SuiteSparse
- OpenLibm
- GMP
- MbedTLS
- LibSSH2
- MPFR
- curl
- LibGit2
- PCRE
- libunwind
- libosxunwind

Furthermore, they default to being used on all systems, as long as it can accurately identify the host system triplet. Use `make USE_BINARYBUILDER=0` to disable.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
I’m very excited about this.

**Comment by staticfloat** [comment]
TODO:

- We've got weird `llvm-config` problems on MacOS and FreeBSD where the BB environment's flags are making it out into `llvm-config --cflags`.  I hypothesize this is because on those platforms `llvm-config` is being built by `clang`, and it keeps more flags than it otherwise would in that case, which we definitely don't want.

- PCRE2 on Windows is calling itself `libpcre2-8-0.dll` instead of the expected `libpcre2-8.dll`.  Fix that.

- `openblas_set_num_threads` threw a fit on Aarch64

- linux64's `libgit2` didn't load its certificates properly; this is probably a problem with my `libgit2` build, I'm missing a patch or something.

File: deps/Versions.make [hunk 1]
```diff
@@ -1,18 +1,32 @@
...
 UNWIND_VER = 1.3.1
 OSXUNWIND_VER = 0.0.5
 GMP_VER = 6.1.2
+GMP_BB_REL = 0
 MPFR_VER = 4.0.2
+MPFR_BB_REL = 0
 PATCHELF_VER = 0.9
 MBEDTLS_VER = 2.16.0
-CURL_VER = 7.56.0
+MBEDTLS_BB_REL = v0.17.0
```
Comment (quinnj) [inline deps/Versions.make]: Just a thumbs up here on the right version here for mbedtls 👍 . That's about all I have to contribute in terms of review!

**Comment by ViralBShah** [comment]
I tried `make` on this branch, but it started building openblas. Perhaps not ready to test just yet?

**Comment by vtjnash** [comment]
The BinaryBuilder files for Win32 SuiteSparse are built incorrectly (https://github.com/JuliaPackaging/Yggdrasil/issues/7)

**Comment by ViralBShah** [comment]
I guess we should get dsfmt into BB as well.

**Comment by StefanKarpinski** [comment]
It doesn’t have to be all or nothing, much easier to get most deps using BB and then add more later.

**Comment by staticfloat** [comment]
Lots of progress; many platforms building properly now; but `MacOS` has started giving this mysterious error during bootstrap, which is very difficult to track down:

```
InexactError(:check_top_bit, UInt32, 0xffffffd0)
```

As I insert `println()` statements to try and figure out where this is coming from, it seems to move around.  My best guess is that it's due to a problem within `PCRE`; some kind of incompatibility, but I haven't been able to figure out where `check_top_bit` is being called from, much less why it's throwing an `InexactError`.

In other news, our fastest platform (`linux64`, running on nanosoldier) did the full build+test in less than 45 minutes, which makes me pretty happy.

**Comment by staticfloat** [comment]
That's a lot of green.  I'm satisfied with this now.  I'm going to do a giant rebase to get to a reasonable number of commits, and then I think we're ready to merge.

**Comment by staticfloat** [comment]
Okay actually it looks like the `libuv` build is causing a segfault on FreeBSD.  I thought this might be because of some missing `-fPIC` flags, but it doesn't look like it.

**Comment by staticfloat** [comment]
Switching this back to WIP until I figure out the libuv thing on FreeBSD

**Comment by staticfloat** [comment]
Wow, green flush!  I think this is ready to merge.

**Comment by fredrikekre** [comment]
I get
```
error during bootstrap:
LoadError("sysimg.jl", 3, LoadError("Base.jl", 363, LoadError("version.jl", 50, InexactError(:check_top_bit, UInt32, 0xffffffd0))))
rec_backtrace at /Users/fredrik/julia-master/src/stackwalk.c:94
record_backtrace at /Users/fredrik/julia-master/src/task.c:210
jl_throw at /Users/fredrik/julia-master/src/task.c:417
jl_parse_eval_all at /Users/fredrik/julia-master/src/ast.c:895
jl_load at /Users/fredrik/julia-master/src/toplevel.c:861
exec_program at /Users/fredrik/julia-master/usr/bin/julia (unknown line)
true_main at /Users/fredrik/julia-master/usr/bin/julia (unknown line)
main at /Users/fredrik/julia-master/usr/bin/julia (unknown line)

*** This error might be fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/fredrik/julia-master/usr/lib/julia/sys.ji] Error 1
make: *** [julia-sysimg-ji] Error 2
```
on a fresh clone.

**Comment by KristofferC** [comment]
Works for me on Windows.

**Comment by ViralBShah** [comment]
I now note that on mac, everything except objconv and dsfmt is coming from BB. It seems that yggdrasil does have objconv and JuliaMath has a dsfmt builder. 

@staticfloat I assume you have this coming in and a new issues isn't necessary.

**Comment by Gnimuc** [comment]
Will the next official Julia minor release use BB2-built binaries as well?

**Comment by staticfloat** [comment]
That's the way things look, yes.

**Pull request merged**